diff --git a/support/PhabricatorStartup.php b/support/PhabricatorStartup.php
--- a/support/PhabricatorStartup.php
+++ b/support/PhabricatorStartup.php
@@ -414,6 +414,9 @@
       switch ($type) {
         case INPUT_SERVER:
           $_SERVER = array_merge($_SERVER, $filtered);
+          if (isset($GLOBALS['PREAMBLE']['SERVER'])) {
+            $_SERVER = array_merge($_SERVER, $GLOBALS['PREAMBLE']['SERVER']);
+          }
           break;
         case INPUT_GET:
           $_GET = array_merge($_GET, $filtered);

